File: procedural/queens.go
package main

import ("fmt"; "os"; "strconv")

func main() {
    size, _ := strconv.Atoi(os.Args[1])
    row := make([]bool, size)
    up := make([]bool, 2 * size - 1)
    down := make([]bool, 2 * size - 1)
    board := make([]int, size)

    var queen func(int)
    queen = func(n int) {
        var i int

        if n >= size {
            for i = 0; i < size - 1; i++ {
                fmt.Printf("%d ", board[i])
            }
            fmt.Println(board[i])
        } else {
            for i = 0; i < size; i++ {
                if !row[i] && !up[n+i] && !down[n-i+size-1] {
                    row[i], up[n+i], down[n-i+size-1] = true, true, true
                    board[n] = i + 1
                    queen(n + 1)
                    row[i], up[n+i], down[n-i+size-1] = false, false, false
                }
            }
        }
    }
    queen(0)
}

File: procedural/queens.c
#include <stdio.h>
#include <stdlib.h>

int size;
int *board;
int *row;
int *up;
int *down;

void queen(int n);

int main(int argc, char *argv[])
{
    if (argc > 1) size = atoi(argv[1]);

    board = calloc(size, sizeof(int));
    row = calloc(size, sizeof(int));
    up = calloc(2*size-1, sizeof(int));
    down = calloc(2*size-1, sizeof(int));

    queen(0);

    return 0;
}

void queen(int n)
{
    int i;

    if (n >= size) {
        for (i = 0; i < size-1; ++i) {
            printf("%d ", board[i]);
        }
        printf("%d\n", board[i]);
    } else {
        for (i = 0; i < size; ++i) {
            if (!row[i] && !up[n+i] && !down[n-i+size-1]) {
                row[i] = up[n+i] = down[n-i+size-1] = 1;
                board[n] = i+1;
                queen(n+1);
                row[i] = up[n+i] = down[n-i+size-1] = 0;
            }
        }
    }
}
